File: packages/android/fernlink-sdk/src/main/kotlin/xyz/fernlink/sdk/SolanaRpc.kt
package xyz.fernlink.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

internal class SolanaRpc(private val endpoint: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    private val json = "application/json".toMediaType()

    data class SignatureStatus(
        val status: TxStatus,
        val slot: Long,
        val blockTime: Long = 0,
    )

    suspend fun getSignatureStatus(signature: String): SignatureStatus = withContext(Dispatchers.IO) {
        val bodyObj = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("id", 1)
            put("method", "getSignatureStatuses")
            put("params", org.json.JSONArray().apply {
                put(org.json.JSONArray().apply { put(signature) })
                put(JSONObject().apply { put("searchTransactionHistory", true) })
            })
        }
        val body = bodyObj.toString().toRequestBody(json)

        val request = Request.Builder().url(endpoint).post(body).build()
        val response = client.newCall(request).execute()
        val text = response.body?.string() ?: throw RuntimeException("empty RPC response")

        val root   = JSONObject(text)
        val result = root.optJSONObject("result") ?: return@withContext SignatureStatus(TxStatus.UNKNOWN, 0)
        val value  = result.optJSONArray("value")?.optJSONObject(0)
            ?: return@withContext SignatureStatus(TxStatus.UNKNOWN, 0)

        val slot   = value.optLong("slot", 0)
        val hasErr = !value.isNull("err")
        SignatureStatus(
            status = if (hasErr) TxStatus.FAILED else TxStatus.CONFIRMED,
            slot   = slot,
        )
    }
}

File: packages/ios/Sources/FernlinkSDK/SolanaRpc.swift
import Foundation

struct SolanaRpc {
    let endpoint: URL

    init(_ urlString: String) {
        endpoint = URL(string: urlString)!
    }

    func getSignatureStatus(_ signature: String) async throws -> SignatureStatus {
        let body: [String: Any] = [
            "jsonrpc": "2.0", "id": 1,
            "method": "getSignatureStatuses",
            "params": [[signature], ["searchTransactionHistory": true]]
        ]
        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        req.timeoutInterval = 10

        let (data, _) = try await URLSession.shared.data(for: req)
        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]

        if let error = json["error"] as? [String: Any],
           let msg   = error["message"] as? String {
            throw FernlinkError.rpcError(msg)
        }

        let result = json["result"] as? [String: Any]
        let values = result?["value"] as? [Any]
        let entry  = values?.first.flatMap { $0 as? [String: Any] }

        guard let entry else {
            return SignatureStatus(status: .unknown, slot: 0, blockTime: 0)
        }

        let slot      = entry["slot"] as? UInt64 ?? 0
        let hasError  = entry["err"] != nil && !(entry["err"] is NSNull)
        return SignatureStatus(
            status:    hasError ? .failed : .confirmed,
            slot:      slot,
            blockTime: 0
        )
    }
}

enum FernlinkError: Error {
    case rpcError(String)
    case proofSigningFailed
    case notStarted
}
